File: libs/moqtail-ts/src/model/control/unsubscribe_namespace.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ByteBuffer, FrozenByteBuffer, BaseByteBuffer } from '../common/byte_buffer'
import { ControlMessageType } from './constant'
import { Tuple } from '../common/tuple'
import { LengthExceedsMaxError } from '../error/error'

export class UnsubscribeNamespace {
  constructor(public readonly trackNamespacePrefix: Tuple) {}

  getType(): ControlMessageType {
    return ControlMessageType.UnsubscribeNamespace
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.UnsubscribeNamespace)
    const payload = new ByteBuffer()
    payload.putTuple(this.trackNamespacePrefix)
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError('UnsubscribeNamespace::serialize(payload_length)', 0xffff, payloadBytes.length)
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): UnsubscribeNamespace {
    const trackNamespacePrefix = buf.getTuple()
    return new UnsubscribeNamespace(trackNamespacePrefix)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest
  describe('UnsubscribeNamespace', () => {
    test('roundtrip', () => {
      const trackNamespacePrefix = Tuple.fromUtf8Path('un/announce/me')
      const unsubscribeNamespace = new UnsubscribeNamespace(trackNamespacePrefix)
      const frozen = unsubscribeNamespace.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.UnsubscribeNamespace))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = UnsubscribeNamespace.parsePayload(frozen)
      expect(deserialized.trackNamespacePrefix.equals(trackNamespacePrefix)).toBe(true)
      expect(frozen.remaining).toBe(0)
    })

    test('excess roundtrip', () => {
      const trackNamespacePrefix = Tuple.fromUtf8Path('un/announce/me')
      const unsubscribeNamespace = new UnsubscribeNamespace(trackNamespacePrefix)
      const serialized = unsubscribeNamespace.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer(excess.length)
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.UnsubscribeNamespace))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const deserialized = UnsubscribeNamespace.parsePayload(frozen)
      expect(deserialized.trackNamespacePrefix).toEqual(trackNamespacePrefix)
      expect(frozen.remaining).toBe(3)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })

    test('partial message', () => {
      const trackNamespacePrefix = Tuple.fromUtf8Path('un/announce/me')
      const unsubscribeNamespace = new UnsubscribeNamespace(trackNamespacePrefix)
      const serialized = unsubscribeNamespace.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        UnsubscribeNamespace.parsePayload(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/control/unsubscribe_namespace.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use super::constant::ControlMessageType;
use super::control_message::ControlMessageTrait;
use crate::model::common::tuple::Tuple;
use crate::model::common::varint::BufMutVarIntExt;
use crate::model::error::ParseError;
use bytes::{BufMut, Bytes, BytesMut};

#[derive(Debug, PartialEq, Clone)]
pub struct UnsubscribeNamespace {
  pub track_namespace_prefix: Tuple,
}

impl UnsubscribeNamespace {
  pub fn new(track_namespace_prefix: Tuple) -> Self {
    Self {
      track_namespace_prefix,
    }
  }
}

impl ControlMessageTrait for UnsubscribeNamespace {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::UnsubscribeNamespace)?;

    let mut payload = BytesMut::new();
    payload.extend_from_slice(&self.track_namespace_prefix.serialize()?);

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "UnsubscribeNamespace::serialize(payload_length)",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;
    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);

    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let track_namespace_prefix = Tuple::deserialize(payload)?;
    Ok(Box::new(UnsubscribeNamespace {
      track_namespace_prefix,
    }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::UnsubscribeNamespace
  }
}
#[cfg(test)]
mod tests {

  use super::*;
  use crate::model::common::varint::BufVarIntExt;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let track_namespace_prefix = Tuple::from_utf8_path("un/announce/me");
    let unsubscribe_namespace = UnsubscribeNamespace {
      track_namespace_prefix,
    };
    let mut buf = unsubscribe_namespace.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::UnsubscribeNamespace as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = UnsubscribeNamespace::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, unsubscribe_namespace);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let track_namespace_prefix = Tuple::from_utf8_path("un/announce/me");
    let unsubscribe_namespace = UnsubscribeNamespace {
      track_namespace_prefix,
    };

    let serialized = unsubscribe_namespace.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::UnsubscribeNamespace as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = UnsubscribeNamespace::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, unsubscribe_namespace);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let track_namespace_prefix = Tuple::from_utf8_path("un/announce/me");
    let unsubscribe_namespace = UnsubscribeNamespace {
      track_namespace_prefix,
    };
    let mut buf = unsubscribe_namespace.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::UnsubscribeNamespace as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = UnsubscribeNamespace::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}
